File: src/main/java/it/gabrieletondi/telldontaskkata/domain/Product.java
package it.gabrieletondi.telldontaskkata.domain;

public class Product {
  private String name;
  private Money price;
  private Category category;

  public String getName() {
    return name;
  }

  public void setName(String name) {
    this.name = name;
  }

  public Money getPriceWithoutTax() {
    return price;
  }

  public void setPriceWithoutTax(Money price) {
    this.price = price;
  }

  public void setCategory(Category category) {
    this.category = category;
  }

  public Money getPriceWithTax() {
    return price.add(getTax());
  }

  public Money getTax() {
    return price.getTax(this.category.getTaxPercentage());
  }
}

File: src/domain/Product.ts
import Category from './Category';

class Product {
  private name: string;
  private price: number;
  private category: Category;

  public getName(): string {
    return this.name;
  }

  public setName(name: string): void {
    this.name = name;
  }

  public getPrice(): number {
    return this.price;
  }

  public setPrice(price: number): void {
    this.price = price;
  }

  public getCategory(): Category {
    return this.category;
  }

  public setCategory(category: Category): void {
    this.category = category;
  }

  public getUnitaryTax(): number {
    return Math.round(this.price / 100 * this.category.getTaxPercentage() * 100) / 100;
  }

  public getUnitaryTaxedAmount(): number {
    return Math.round((this.price + this.getUnitaryTax()) * 100) / 100;
  }
}

export default Product;

